### ISSUE TITLE
Enhance ML Model & Training Quality

### ISSUE THREAD

**Issue by solkem** [solkem/edgechain-midnight-hackathon#8]
## Description

Improve the machine learning model architecture, training quality, and add proper evaluation metrics for the federated learning system.

## Current State

✅ Basic TensorFlow.js training working
✅ Synthetic crop data generation
✅ Model weight extraction
❌ Simple model architecture (may not be optimal)
❌ No model validation/testing
❌ No accuracy tracking across rounds

### 1. Improve Model Architecture

Current model in `packages/ui/src/lib/federatedLearning.ts`:

- [ ] Research crop yield prediction models
- [ ] Implement more sophisticated architecture
- [ ] Add dropout layers for regularization
- [ ] Tune hyperparameters (learning rate, batch size)
- [ ] Consider ensemble approaches

### 2. Better Training Data

Improve `generateSyntheticData()`:

- [ ] Use realistic crop yield data ranges
- [ ] Add more features:
  - Soil pH
  - Fertilizer usage
  - Irrigation
  - Crop rotation
  - Pest/disease history
- [ ] Add data augmentation
- [ ] Support multiple crop types (maize, wheat, rice)
- [ ] Use real historical data if available

### 3. Model Evaluation

Add evaluation metrics:

- [ ] Split data into train/validation/test sets
- [ ] Track validation loss
- [ ] Calculate R² score
- [ ] Mean Absolute Error (MAE)
- [ ] Root Mean Square Error (RMSE)
- [ ] Display metrics in UI

### 4. Cross-Round Performance Tracking

- [ ] Store model performance history
- [ ] Compare current model vs previous rounds
- [ ] Visualize accuracy improvement over time
- [ ] Detect model degradation
- [ ] Alert if aggregation makes model worse

### 5. Differential Privacy

Add privacy guarantees:

- [ ] Implement differential privacy in training
- [ ] Add Gaussian noise to gradients
- [ ] Calculate privacy budget (ε, δ)
- [ ] Display privacy guarantees to users
- [ ] Document privacy-utility tradeoffs

### 6. Model Compression

Optimize for resource-constrained farmers:

- [ ] Implement weight quantization
- [ ] Reduce model size
- [ ] Faster inference for predictions
- [ ] Test on low-end devices/browsers

### 7. Testing & Benchmarking

Create `packages/ui/src/lib/__tests__/federatedLearning.test.ts`:

- [ ] Unit tests for training functions
- [ ] Test model serialization/deserialization
- [ ] Benchmark training time
- [ ] Test with different data sizes
- [ ] Test convergence behavior

## Acceptance Criteria

- [ ] Model achieves reasonable accuracy (>80% on test set)
- [ ] Training completes in < 2 minutes on average hardware
- [ ] Validation metrics displayed in UI
- [ ] Model performance improves across rounds
- [ ] Differential privacy implemented with documented guarantees
- [ ] Model size < 5MB for fast downloads
- [ ] Comprehensive tests with >80% coverage
- [ ] Documentation of model architecture and training process

## Related Files

- `packages/ui/src/lib/federatedLearning.ts` - Main training logic
- `packages/ui/src/components/FLDashboard.tsx` - UI for metrics

## Technical Resources

- [TensorFlow.js Models](https://www.tensorflow.js.org/tutorials)
- [Federated Learning Best Practices](https://arxiv.org/abs/1912.04977)
- [Differential Privacy in FL](https://arxiv.org/abs/1710.06963)
- [Crop Yield Prediction](https://www.kaggle.com/datasets)

## Example Improved Architecture

```typescript
const model = tf.sequential({
  layers: [
    tf.layers.dense({ inputShape: [10], units: 64, activation: 'relu' }),
    tf.layers.dropout({ rate: 0.2 }),
    tf.layers.dense({ units: 32, activation: 'relu' }),
    tf.layers.dropout({ rate: 0.2 }),
    tf.layers.dense({ units: 16, activation: 'relu' }),
    tf.layers.dense({ units: 1 }) // Yield prediction
  ]
});

model.compile({
  optimizer: tf.train.adam(0.001),
  loss: 'meanSquaredError',
  metrics: ['mae', 'mse']
});
```

## Dependencies

- Can be implemented in parallel with other issues
- Improves overall system quality
